### ISSUE TITLE
Incorrect join partition elimination

### ISSUE THREAD

**Issue by darthunix** [greenplum-db/gporca#565]
When we join a partitioned table with a non-partitioned one join partition elimination doesn't exclude all partitions we expect. 
```sql
create table t1 (a int, b timestamp)
distributed by (a)
partition by range (b)
subpartition by list (a)
subpartition template
(
    subpartition sp_0 values (0),
    subpartition sp_1 values (1)
)
(
    partition dec_2019 start (timestamp '2019-12-01 00:00:00') inclusive,
    partition jan_2020 start (timestamp '2020-01-01 00:00:00') inclusive
        end (timestamp '2020-02-01 00:00:00') exclusive
);

create table t2 (a int) distributed by (a);
insert into t2 values (1);

explain analyze
select * from t1 join t2
on t1.a = t2.a and t1.b = '2019-12-02 00:00:00';
```
Generated plan eliminates only 50% of partitions (`a = 1` filter) and scans two of them while the proper behaviour is to scan exactly one partition ( `a = 1 and b = '2019-12-02 00:00:00'`).
```
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=1 width=16) (actual time=14.916..14.916 rows=0 loops=1)
   ->  Hash Join  (cost=0.00..862.00 rows=1 width=16) (never executed)
         Hash Cond: (t1.a = t2.a)
         ->  Dynamic Seq Scan on t1 (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=12) (never executed)
               Filter: (b = '2019-12-02 00:00:00'::timestamp without time zone)
               Partitions scanned:  2 (out of 4)  (seg1).
         ->  Hash  (cost=100.00..100.00 rows=34 width=4) (actual time=0.694..0.694 rows=1 loops=1)
               ->  Partition Selector for t1 (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4) (actual time=0.679..0.681 rows=1 loops=1)
                     ->  Seq Scan on t2  (cost=0.00..431.00 rows=1 width=4) (actual time=0.167..0.168 rows=1 loops=1)
```
The root of the problem seems to be the content of PartEqFilters in PartitionSelector node. A ConstValue is `true` while it is expected to be `2019-12-02 00:00:00` - so partition elimination logic for PartitionSelector node doesn't have enough information to exclude all garbage subpartitions.
```
<dxl:PartEqFilters>
    <dxl:ConstValue TypeMdid="0.16.1.0" Value="true"/>
    <dxl:Ident ColId="9" ColName="a" TypeMdid="0.23.1.0"/>
</dxl:PartEqFilters>
```
For example in a case of a simple query to `t1` without any joins PartEqFilters contains correct elimination constants
```sql
explain analyze verbose 
select * from t1 where a = 1 and b = '2019-12-02 00:00:00';
```
```
<dxl:PartEqFilters>
    <dxl:ConstValue TypeMdid="0.1114.1.0" Value="ACCFB6w7AgA=" DoubleValue="628560000000000.000000"/>
    <dxl:ConstValue TypeMdid="0.23.1.0" Value="1"/>
</dxl:PartEqFilters>
```
So we get exactly one partition as expected
```
Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.00 rows=1 width=12) (actual time=1.275..1.275 rows=0 loops=1)
->  Sequence  (cost=0.00..431.00 rows=1 width=12) (never executed)
      ->  Partition Selector for t1 (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4) (never executed)
            Partitions selected: 1 (out of 4)
      ->  Dynamic Seq Scan on t1 (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=12) (never executed)
            Filter: ((a = 1) AND (b = '2019-12-02 00:00:00'::timestamp without time zone))
            Partitions scanned:  1 (out of 4) .
```
I think that a problem is in `CPartitionPropagationSpec::SplitPartPredicates` or `CPartitionPropagationSpec::AppendEnforcers` functions of a `CPartitionPropagationSpec.cpp` file. But I can't exactly understand how they work in details to make a PR (may be someone can help me and give any explanations?).
Also I attach ORCA mini dump files with [join t1 and t2](https://github.com/greenplum-db/gporca/files/4068598/PartitionEliminationJoinConst.mdp.txt) and a [simple t1 partition filter](https://github.com/greenplum-db/gporca/files/4068602/PartitionEliminationSimple.mdp.txt)

**Comment by zellerh** [comment]
Thanks very much for reporting this, @darthunix. We have seen similar queries and agree that this would be something we could improve.

As far as I understand, the issue is whether we can create partition selector(s) for a dynamic table scan that combine a local predicate on the partitioned table and a predicate that comes from another table.

Trying a few simple test cases makes me think that we currently can use only one partition selector for each dynamic table scan, but I could be wrong.

@asubramanya suggested this SIGMOD article on partition selection: https://dl.acm.org/doi/10.1145/2588555.2595640. If needed we can also provide a PDF.

Unfortunately we don't have a simple suggestion for how to implement this enhancement yet. If you are interested in contributing a fix we can think about it some more.

**Comment by darthunix** [comment]
Thanks for an article, @zellerh and @asubramanya - finally I have read it. I can try to take this improvement as an afterwork activity (to understand ORCA better). So I can't promise, that I will write a PR quickly (it seems to be a big work).

I don't know ORCA's internals, so may be you can advise a high level the architecture of the patch?
From the first point of view it seems that we need to allow a DynamicScan node (consumer) subscribe to several PartitionSelector nodes (producers) - not to exact one as now. But it requires intersection  for partition OIDs from many PartitionSelector nodes in some buffer (DynamicScan node or we need to implement some other node type). But may be there is a more simple and explicit solution.

**Comment by d** [comment]
Hi Denis,
Thanks for bringing this to our attention. I talked to Hans offline, posting here for posterity:

TL;DR: Dynamic Table Scan currently selects the _union_ of partitions selected all partition selectors, instead of the _intersection_ of them.

Long version:
Intuitively, it takes two partition selectors to implement the feature you request: one hardcodes the partitions _statically_ selected, the other one dynamically selects partitions at runtime. *But* the executor currently does not have the intelligence to discard partitions that are selected by only a subset of the partition selectors. This requires work on both the optimizer and the executor. A couple of us have ventured onto a quick spike in the summer of 2017 and had a patch, but it broke another feature of partial indexes over partition tables so we set the work down and left it to the PM to re-prioritize.

Ciao,
Jesse
